src/aiortc/contrib/media.py

```diff
@@ -530,18 +535,23 @@ async def __run_track(self, track, context):
             else:
                 # keypoint stream
                 received_keypoints = frame.data
+                asyncio.run_coroutine_threadsafe(self.keypoints_queue.put(received_keypoints), loop)
                 self.__log_debug("Keypoints for frame %s received at time %s",
                                 str(self.received_keypoints_frame), time.time())
 
-                keypoints_file = open(self.__keypoints_file_name, "a")
-                keypoints_file.write(str(received_keypoints))
-                keypoints_file.write("\n")
-                keypoints_file.close()
+                if save_keypoints_to_file:
+                    keypoints_file = open(self.__keypoints_file_name, "a")
+                    keypoints_file.write(str(received_keypoints))
+                    keypoints_file.write("\n")
+                    keypoints_file.close()
 
-                if enable_prediction:
+                if enable_prediction and not self.video_queue.empty():
                     try:
+                        received_keypoints = await self.keypoints_queue.get()
+
                         before_predict_time = time.time()
-                        predicted_target = model.predict(received_keypoints)
+                        with concurrent.futures.ThreadPoolExecutor() as pool:
```

Comment:
Can you point me to a reference for this? Wondering if this is the best way to call a function asynchronously, but not entirely sure what other options are. What went into the decision to use this?

Reply:
What I found was that although __run_track is supposed to be a non-blocking event loop, it could fail when there is a blocking code in it. So I looked for how to run a blocking io code in a non-blocking way in event loops. I found this guide that enables a blocking io to run in a custom thread pool: https://docs.python.org/3/library/asyncio-eventloop.html#executing-code-in-thread-or-process-pools